Added MAVLink 2.0 message signing and authentication support to provide secure communication between GCS and flight controller.

PritamP20 opened the pull request:
## Summary

   This PR adds MAVLink 2.0 message signing and authentication support to provide secure communication between the ground control station and flight controller.

   ## Motivation

   MAVLink signing is essential for:
   - **Security**: Protect against unauthorized parameter changes and message injection
   - **Compliance**: Required for commercial drone operations in many jurisdictions
   - **Safety**: Prevent malicious actors from taking control of vehicles

   ## Changes

   ### New Files
   - `ardupilot_methodic_configurator/backend_signing_keystore.py` (450 lines)
     - Secure key storage using OS keyring with encrypted file fallback
     - Key generation, storage, retrieval, deletion
     - Password-protected import/export
   
   - `ardupilot_methodic_configurator/data_model_signing_config.py` (150 lines)
     - Configuration data model with validation
     - Serialization/deserialization
   
   - `tests/test_signing_keystore.py` (15 test cases)
   - `tests/test_signing_config.py` (15 test cases)
   - `docs/mavlink_signing/` (complete documentation)
   - `install_signing_dependencies.sh` - Installation script

   ### Modified Files
   - `ardupilot_methodic_configurator/backend_flightcontroller.py`
     - Added `setup_signing()` method
     - Added `_send_setup_signing_command()` method
     - Added `_unsigned_callback()` method
     - Added `get_signing_status()` method
   
   - `pyproject.toml`
     - Added `cryptography>=41.0.0` dependency
     - Added `keyring>=24.0.0` dependency

   ## Features

   ### Security
   - ✅ HMAC-SHA-256 message signing (authentication, not encryption)
   - ✅ OS keyring integration (Windows/macOS/Linux)
   - ✅ AES-256 encrypted file fallback
   - ✅ Cryptographically secure key generation
   - ✅ Password-protected key export/import
   - ✅ Per-vehicle key isolation
   - ✅ Timestamp-based replay protection

   ### Key Management
   - ✅ Generate cryptographically secure 32-byte keys
   - ✅ Store keys securely per vehicle
   - ✅ Retrieve keys for signing setup
   - ✅ Delete keys when no longer needed
   - ✅ List all vehicles with configured keys
   - ✅ Export/import keys with password protection

amilcarlucas commented:
Thanks, thiswill take a while to review. Have you tested this in multiple vehicles?

PritamP20 commented:
Yes, I’ve tested it with multiple vehicle entries using the **test_multiple_vehicles** and **test_list_vehicles_with_keys** unit tests. These tests cover storing, retrieving, and managing keys for different vehicle IDs to ensure proper isolation and persistence between them.

PritamP20 commented:
yup working on it @amilcarlucas 
I will update the pr soon

amilcarlucas commented:
Any progress on this?

PritamP20 commented:
@amilcarlucas I tried running it —
./scripts/run_sitl_tests.sh download works fine, but when I run
```
./scripts/run_sitl_tests.sh run
```

I get an error because the script has hardcoded Linux paths in the download_sitl() function (around line 112). It doesn’t detect macOS properly, so it ends up trying to download the Linux binaries instead.

I’m currently working on adapting the script for macOS. If that doesn’t work out, I’ll look for an alternative way to run the tests.

PritamP20 commented:
Thank you @amilcarlucas it works, i have dockerised the project and the SITL tests are working and one test is failing i will fix it and update the branch

PritamP20 commented:
All test are passing!!!
I have wrote the dockerfile and executing the SITL test in linux environment

<img width="1245" height="717" alt="Screenshot 2026-01-13 at 4 51 28 PM" src="https://github.com/user-attachments/assets/2101c60d-eaf0-44ec-8d2b-5a6e4978b5b8" />

PritamP20 commented:
I am using vehicle ID because it helps uniquely identify each drone’s signing key, this ensures key persistence across sessions and isolates keys for multi-vehicle setups, so one compromised key doesn’t affect others

amilcarlucas commented:
Please document that in the code.

amilcarlucas commented on ardupilot_methodic_configurator/backend_signing_keystore.py:
```diff
@@ -0,0 +1,548 @@
+            if exc_name == "NoKeyringError":
+                self._logger.debug("No keyring backend available")
+            else:
+                self._logger.debug("Keyring test failed: %s", exc)
```
Again, warnings, not debug

amilcarlucas commented on tests/bdd_signing_config.py:
needs to be renamed with
git mv tests/bdd_signing_config.py tests/bdd_data_model_signing_config.py

amilcarlucas commented on tests/bdd_signing_keystore.py:
git mv tests/bdd_signing_keystore.py tests/bdd_backend_signing_keystore.py

amilcarlucas reviewed (changes requested):
This is getting better

PritamP20 commented:
Thank you for the suggestions @amilcarlucas i will fix changes requested

PritamP20 commented on ardupilot_methodic_configurator/backend_signing_keystore.py:
```diff
@@ -0,0 +1,548 @@
+    Secure keystore for MAVLink 2.0 signing keys.
+
+    This class provides secure storage for signing keys using:
+    1. OS keyring (preferred) - Windows Credential Manager, macOS Keychain, Linux Secret Service
```
Okay i will put this in the contribution docs

PritamP20 commented on ardupilot_methodic_configurator/backend_signing_keystore.py:
```diff
@@ -0,0 +1,548 @@
+                self._logger.warning("Failed to cleanup keyring test: %s", cleanup_exc)
+            return True
+        except ImportError:
+            self._logger.debug("Keyring package not installed")
```
Got it!!

PritamP20 commented:
@amilcarlucas can you please review it

PritamP20 commented:
@amilcarlucas it's okay, take your time till then i will work on dockerise the project for developer friendly testing and development

amilcarlucas commented on ardupilot_methodic_configurator/backend_flightcontroller.py:
```diff
@@ -475,6 +477,190 @@ def is_battery_monitoring_enabled(self) -> bool:
         """Check if battery monitoring is enabled - delegates to commands manager."""
         return self._commands_manager.is_battery_monitoring_enabled()
 
+    # MAVLink Signing Functionality
```
Please add BDD tests for these functions in a new tests/bdd_backend_flightcontroller_signing.py file.

PritamP20 commented:
@amilcarlucas Thank you Reviewing it, I will make the requested changes

amilcarlucas commented on tests/unit_signing_keystore.py:
```diff
@@ -0,0 +1,343 @@
+#!/usr/bin/env python3
```
git mv tests/unit_signing_keystore.py tests/unit_backend_signing_keystore.py

PritamP20 reviewed (commented):
I have done all of them

amilcarlucas reviewed (changes requested):
Thanks, It is a lot better now.

amilcarlucas committed:
test(security): Add tests for the MAVLink signing functionality

PritamP20 commented:
@amilcarlucas sure i will check them and get back to you

amilcarlucas commented:
I mean tests or real hardware, to be 100% sure it works before merging 3800 lines of code.

PritamP20 commented:
@amilcarlucas I’ll look into the requested changes. By the way, I’m applying for GSoC 2026 with ArduPilot  would you be open to mentoring me? I’d love to continue working on this feature and contributing to this repository.

amilcarlucas commented:
Yes, I will gladly mentor you.

amilcarlucas committed:
feat(security): add MAVLink 2.0 message signing and authentication

Implement secure communication between ground control station and flight
controller using HMAC-SHA256 message signing to prevent tampering and
message injection attacks.

New Components:
- backend_flightcontroller: setup_signing(), disable_signing(),
  get_signing_status() methods for MAVLink signing management
- backend_signing_config: configuration persistence with atomic file
  operations and cross-platform locking
- backend_signing_keystore: secure key storage using OS keyring with
  encrypted file fallback

Security Features:
- 32-byte (256-bit) cryptographic signing keys
- Timestamp-based replay attack prevention (7-day validation window)
- OS-native keyring support (Windows Credential Manager, macOS Keychain,
  Linux Secret Service)
- Secure file fallback with AES-256 encryption and restrictive permissions
  (0o600 on Unix)
- Per-vehicle key isolation

Documentation:
- Added comprehensive OS keyring setup guide in INSTALL.md
- Security warnings about file-based fallback limitations
- Platform-specific installation instructions

Implementation Details:
- Configurable signing modes (sign_outgoing, allow_unsigned_in)
- Permissive unsigned message handling during setup phase
- Specific exception handling for file operations
- Type-safe implementation with full type hints
- Extensive validation (key length, link_id range, timestamp age)

Related: #1038

Signed-off-by: PritamP20 <pripritam7@gmail.com>

amilcarlucas committed:
fix(MAVLink signing): Removed filesystem key storage feature

It is unsafe, and adds complexity.

amilcarlucas committed:
test(MAVLink signing): Removed tests for removed feature

amilcarlucas commented:
It's rebased, and the filesystem fallback code removed, and even more tests added.
@PritamP20 Are you going to continue working on this to add a GUI ?

amilcarlucas committed:
test(keyring): Add keyring tests for all OS

amilcarlucas commented:
@PritamP20 did you abandoned this?
